Alternative pin group based pinctrl implementation for RT1062

danieldegrasse opened the pull request:
This PR is an alternative approach from #41810 which uses pin groups instead of pin nodes defined at the SOC level. An SOC level header file is still included for readability and so that the user is aware of all pinmux options present on the SOC.

The pin groups approach offers the following benefits:
- multiple pinctrl nodes can be created with the same pinmux value, enabling pins to have a different states in different power modes without changing their mux settings
- selecting a pinctrl group in the board level DTS is simpler and less error prone than individually selecting pins for that peripheral

hakehuang committed:
dts: bindings: add pinctrl binding for nxp mcux rt1xxx

Add dts binding for rt1xxx pinctrl driver settings. A binding file is
present for the pinctrl node itself, and for the pinctrl child node that
defines all pinmux options

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
dts: bindings: add pinctrl node to sai and lpuart

enable pinctrl in i2s and lpuart driver bindings

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
soc: add pinctrl header file definition for RT series

Pinctrl requires header file with Z_PINCTRL_STATE_PINS_INIT macro
defined. Add header file for mcux RT pinctrl implementation.
Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
dts: imx_rt: add pinctrl and gpr for rt1xxx

add dtsi settings for rt series
dtsi use gpr to replace pinmux
nxp iomuxc has gpr which has more settings than mux and io settings
current solution is to export gpr separately and access then directly

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
board: pinctrl: board config for RT1060

enable pin control for RT1060 EVK.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
drivers: pinctrl: add mcux_rt pinctrl driver

Add pinctrl driver for rt1xxx

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>
Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang committed:
drivers: uart: add pinctrl support in mcux lpuart

enable pinctrl in lpuart driver, and clean up driver instance definition
macros

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
drivers: sai: add pinctrl support in mcux sai

enable i2s pinctrl

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

hakehuang committed:
dts: binding: rename pinmux to gpr

rename pinmux to gpr
different from pinmux and io settings gpr will do more IO
settings.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

danieldegrasse committed:
west.yml: Update to pull in pinctrl header from NXP HAL

NXP hal will define constants for pinmux options in RT pinctrl
implementation. Update hal revision to pull in dtsi file.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang commented on drivers/pinctrl/pinctrl_mcux_rt.c:
```diff
@@ -0,0 +1,87 @@
+			input_daisy, (uint32_t)config_register,
+			MCUX_RT_INPUT_ENABLE(pin_ctrl_flags));
+
+		if (MCUX_RT_INPUT_SCHMITT_ENABLE(pin_ctrl_flags)) {
```
I think below judgement is not necessary, as you have shift the pin_ctrl_flags in pinctrl_soc_mcux_rt_common.h, where the dts setting is translated to.

danieldegrasse replied:
Those shifts are chosen arbitrarily, and don't correspond to the actual bitfield location in the control register. If I was careful when selecting the shift offset of these one bit fields, then I could use the value in `pin_ctrl_flags` directly. Even with this, some of the `if` statements are required, because for some fields there is a special "none" value that tells the driver not to configure the property for that pin and leave it at the SOC default.

That being said, for these one bit fields the `if` statement is not necessary, and can be replaced with something like the following:
```
config_val |= IOMUXC_SW_PAD_CTL_PAD_HYS(MCUX_RT_INPUT_SCHMITT_ENABLE(pin_ctrl_flags));
```
Thanks for pointing that out, I'll update the PR accordingly

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,132 @@
+          101 DSE_5_R0_5 — 32 Ohm @3.3V, 52 Ohm @1.8V
+          110 DSE_6_R0_6 — 32 Ohm @3.3V, 43 Ohm @1.8V
+          111 DSE_7_R0_7 — 26 Ohm @3.3V, 37 Ohm @1.8V
+      bias-pull-up:
+        required: false
+        type: int
+        enum:
+          - 0
+          - 1
+          - 2
+          - 3
+        description: |
+          Select the value of the pull up resistor present on this pin
+          Corresponds to the PUS field in the IOMUXC peripheral
+          00 Unused- no change will be applied to pin
+          01 PUS_1_47K_Ohm_Pull_Up — 47K Ohm Pull Up
+          10 PUS_2_100K_Ohm_Pull_Up — 100K Ohm Pull Up
+          11 PUS_3_22K_Ohm_Pull_Up — 22K Ohm Pull Up
```
I'd do here instead (to follow standard properties as much as possible):

```
      # note: bias-pull-up added to the allowlist
      bias-pull-up-value:
        required: false
        type: int
        enum:
          - "unused"
          - "47k"
          - "100k"
          - "22k"
        description: |
          Select the value of the pull up resistor.
```

Then do this in boards:

```
bias-pull-up;
bias-pull-up-value = "100k";
```

and to store value:

```
.pullup = DT_PROP(node_id, bias_pull_up) * DT_ENUM_IDX(node_id, bias_pull_up_value),
```

---

Or maybe:
```c
// include/dt-bindings/pinctrl/nxp/mcux-rt-pinctrl.h

#define MCUX_PULL_47K 0x1
#define MCUX_PULL_100K 0x2
...
```

```
      bias-pull-up-value:
        required: false
        type: int
        description: |
          Select the value of the pull up resistor (use MCUX_PULL_XXX pre-defined vaues).
```

```
bias-pull-up;
bias-pull-up-value = <MCUX_PULL_100K>;
```

```
.pullup = DT_PROP(node_id, bias_pull_up) * DT_PROP(node_id, bias_pull_up_value),
```

danieldegrasse replied:
This makes sense. I'll switch the pull up implementation to use a format like this, and keep the `bias_pull_up` property standard.

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi:
```diff
@@ -0,0 +1,59 @@
+	};
+
+	sai1_default: sai1_default {
+		iomuxc_gpio_ad_b1_09_sai1_mclk: iomuxc_gpio_ad_b1_09_sai1_mclk {
+			pinmux = <IOMUXC_GPIO_AD_B1_09_SAI1_MCLK>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_12_sai1_rx_data00: iomuxc_gpio_ad_b1_12_sai1_rx_data00 {
+			pinmux = <IOMUXC_GPIO_AD_B1_12_SAI1_RX_DATA00>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_13_sai1_tx_data00: iomuxc_gpio_ad_b1_13_sai1_tx_data00 {
+			pinmux = <IOMUXC_GPIO_AD_B1_13_SAI1_TX_DATA00>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_14_sai1_tx_bclk: iomuxc_gpio_ad_b1_14_sai1_tx_bclk {
+			pinmux = <IOMUXC_GPIO_AD_B1_14_SAI1_TX_BCLK>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_15_sai1_tx_sync: iomuxc_gpio_ad_b1_15_sai1_tx_sync {
+			pinmux = <IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC>;
+			bias-bus-hold;
+			input-enable;
+		};
```
you're not applying grouping principle here, you should have:

```
sai1_default: sai1_default {
	pinmux = <IOMUXC_GPIO_AD_B1_09_SAI1_MCLK>,
		        <IOMUXC_GPIO_AD_B1_12_SAI1_RX_DATA00>,
		        <IOMUXC_GPIO_AD_B1_13_SAI1_TX_DATA00>,
		        <IOMUXC_GPIO_AD_B1_14_SAI1_TX_BCLK>,
		        <IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC>;
	bias-bus-hold;
	input-enable;
};
```

danieldegrasse replied:
I don't think it it possible to apply grouping for iMX.RT parts using the pinctrl API. Each pinmux option is defined in a header file as something like this:
`#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC 0x401f8138 3 0x401f85ac 1 0x401f8328`

Those values correspond to the following registers in the datasheet:
- Pin specific `SW_MUX_CTL_PAD` register (0x401f8138 in this example)
- Pin specific value to write to `MUX_MODE` field of the `SW_MUX_CTL_PAD` register (3)
- Peripheral specific `SELECT_INPUT` register (0x401f85ac)
- Value to write to `DAISY` field of `SELECT_INPUT` register (1)
- Pin specific `SW_PAD_CTL_PAD` register (0x401f8328)

All of these values are unique to either the pin or peripheral being selected, so they can't be shared. The issue I run into with pin grouping is that DTS does not have a way to define an "array of arrays", which is what the pinmux option in pin grouping is. I haven't been able to find a workaround for this that does not require modifying the pinctrl API.

Since the maximum offset from the base address of `0x401f8000` requires 10 bits to store, there is no way to pack three register offsets, plus two 3 bit register settings, into a 32 bit integer, the way that implementations like #39974 do.

The only solutions I see so far are these two:
- Do not support pin grouping
- Encode each pinmux option as an array of `phandles`, and store the pinmux data itself in the dts node the `phandle` references.

Do you have any feedback on which option would work better, or is there a trick I'm missing? I'd prefer to use pin grouping, but I'm unsure how to enable it given this issue

gmarull replied:
Have you considered encoding the offsets instead, maybe:

```c
/* Encode:
 * bits 0..11: Offset, used to resolve SW_MUX_CTL_PAD_* and SW_PAD_CTL_PAD_*
 *                  Looks like SW_PAD_CTL_PAD_* = SW_MUX_CTL_PAD_ + 0x200U
 * bits 12..23: 
 * bits 24..26: MUX_MODE (0..7)
 * bit 27: DAISY
 */
#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC (0x138U | (0x5acU << 12U) | (3U << 24U) | (1U << 27U)
```

I'm not familiar enough with your platform to judge if that would scale to all your SoCs.

gmarull replied:
You could also look at encoding _distances_, for example:

Any `IOMUX_SW_MUX_*` ranges from 0x14 to 0x200, 9 bits
Any `IOMUX_SW_PAD_*` ranges from 0x204 to 0x3f0, 9 bits
Any `IOMUX_*_SELECT_INPUT` ranges from 0x3f4 to 0x78c, 10 bits

That leaves 4 bits, 3 for MUX and 1 for DAISY. Again, it's really pushing the limits and may not be exportable to all your SoCs...

gmarull replied:
Another trick I'm thinking of: Encode offsets in words, not bytes.

`IOMUX_SW_MUX_*` can be stored as: (offset - 0x14) / 4. Range of values to save: 0...0x7b (7 bits)
`IOMUX_SW_PAD_*` can be stored as: (offset - 0x204) / 4. Range of values to save: 0...0x7b (7 bits)
`IOMUX_*_SELECT_INPUT` can be stored as: (offset - 0x3f4) / 4. Range of values to save: 0...0x398 (10 bits)

So you have now: 7 + 7 + 10 + 3 + 1 = 28 bits, you are safe to even save some bits.

Then we'd have something like:
```c
#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC (0x49U | (0x49U << 7U) | (0xEA << 14U) | (3U << 24U) | (1U << 27U)
```

danieldegrasse replied:
This could work. The `0x14` and `0x204` offsets can be stored in the DTS for the `pinctrl` itself, so those being SOC specific would be a nonissue. My only concern is that some pins are configured via a different peripheral (`IOMUXC_SVNS`), with a different base address. I'll have to figure out a way to indicate that difference to the driver, but it shouldn't take more than a few bits.

I'll look at reference manuals for the remainder of the iMX.RT line and see if this would work for higher pin count SOCs, and if so implement it. Thanks for the idea!

danieldegrasse commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,132 @@
+          1 SRE_1_Fast_Slew_Rate — Fast Slew Rate
+      nxp,speed:
+        required: false
+        type: int
+        enum:
+          - 0
+          - 1
+          - 2
+          - 3
+        description: |
+          Sets pin speed. Corresponds to SPEED field in IOMUXC peripheral
+          00 SPEED_0_low_50MHz_ — low(50MHz)
+          01 SPEED_1_medium_100MHz_ — medium(100MHz)
+          10 SPEED_2_medium_100MHz_ — medium(100MHz)
+          11 SPEED_3_max_200MHz_ — max(200MHz)
```
I wasn't aware you could do this with an `enum` in devicetree. I'll switch to using human readable values like these where possible. On some pins that second `medium-100mhz` value actually sets the speed to 150MHz, so I'll update that.

danieldegrasse commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,132 @@
+          0 SRE_0_Slow_Slew_Rate — Slow Slew Rate
+          1 SRE_1_Fast_Slew_Rate — Fast Slew Rate
+      nxp,speed:
+        required: false
```
I'm not sure what you mean here. I left all pin properties as `required: false` because the pinctrl driver uses `DT_PROP_OR` to default all these values to a special "none" value if they are not set explicitly in the devicetree. This allows the driver to skip configuring any bit fields in the pin configuration register that the user does not explicitly set with a devicetree property.

So if `nxp,speed` is set in the devicetree, the driver will pick that up and write the provided value to the pin configuration register. Otherwise it will leave the register at its default value.

danieldegrasse commented on soc/arm/nxp_imx/rt/pinctrl_soc.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef ZEPHYR_SOC_ARM_NXP_IMX_RT_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_ARM_NXP_IMX_RT_PINCTRL_SOC_H_
+
+#include <drivers/pinctrl/pinctrl_soc_mcux_rt_common.h>
```
Only will be used by RT series SOCs. I can rename the file.

danieldegrasse commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -335,8 +338,20 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
 	}
+}
 
+#ifdef CONFIG_PINCTRL
+static void mcux_rt_sai_pin_init(void)
+{
+	int err;
+
+	err = pinctrl_apply_state(dev_cfg->pinctrl, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		LOG_ERR("mclk pinctrl setup failed (%d)", err);
+		return err;
+	}
 }
+#endif
```
@hakehuang, could you clarify the reasoning for the helper here?

danieldegrasse commented:
@gmarull I've updated the PR to use string `enums`, per your suggestion. For the pin mux properties, We've prefer to avoid trying to pack all the data into a 32 bit field, primarily for scalability concerns. Some of our higher pin count Linux cores have a companion M4 core, and use the same IOMUXC peripheral. The approach I used in this update was to split pin configuration fields across several properties. The other option I considered was a method using DTS phandles to define the pin mux options, and using those phandles in pin groups, like so:

soc `dtsi` file:
```
&iomuxc {
        iomuxc_gpio_ad_b0_12_lpuart1_tx: IOMUXC_GPIO_AD_B0_12_LPUART1_TX {
        	pinmux = <0x401f80ec 2 0x0 0 0x401f82dc>;
        };
        
        ...
        
        iomuxc_gpio_ad_b0_13_lpuart1_rx: IOMUXC_GPIO_AD_B0_13_LPUART1_RX {
        	pinmux = <0x401f80f0 2 0x0 0 0x401f82e0>;
        };
};
```
board `dtsi` file
```
&pinctrl {
    lpuart1_default: lpuart1_default {
		group0 {
			pinmux = <&iomuxc_gpio_ad_b0_12_lpuart1_tx>,
				<&iomuxc_gpio_ad_b0_13_lpuart1_rx>;
			nxp,pull = "pull-enable";
			bias-bus-hold;
			drive-strength = "r0-6";
			slew-rate = "slow";
			nxp,speed = "mhz-100";
		};
	};
}
```
My primary concern with this approach is that it mixes both pin node and pin group approaches, and a user might incorrectly assume they could use the `iomuxc` nodes as pinctrl nodes directly.

The other change in the latest push is a new property, `nxp,pull`. This property sets the `PKE` bit in the `IOMUXC` peripheral. The rationale was that if a user did not provide either `bias-bus-hold`, `bias-pull-up`, or `bias-pull-down` as a property, the pin would have all pull up and keeper functions disabled. Although this is a valid configuration, the typical "default" configuration for a pin that is not being pulled up on an RT SOC would be to enable the keeper.

Adding this property forces a user to understand that if they don't supply bus bias configuration to the pinctrl driver the pin will have its keeper disabled. The other option I considered here is to default to `PKE=1, PUE=0` if the user does not supply `bias-disable` as a property

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -91,42 +126,54 @@ child-binding:
-          - 2
-          - 3
+          - "unused"
+          - "pull-up-47k-ohm"
+          - "pull-up-100k-ohm"
+          - "pull-up-22k-ohm"
```
suggestion: I'd make enum str shorter, like `47k` or `47kohm`.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -91,42 +126,54 @@ child-binding:
+        description: |
+          Select the value of the pull up resistor present on this pin
+          Corresponds to the PUS field in the IOMUXC peripheral
+          00 PUS_0_100K_Ohm_Pull_Down - 100K Ohm Pull Down
```
since there's only 1 pull-down value, is this required?

danieldegrasse replied:
No, it is not. I added it for the sake of consistency between pull up and pull down resistors, and because in my mind having that value visible in the generated DTS is useful. With the value included, a user can look at their output `zephyr.dts` and see the exact pull up or pull down resistor they have selected. It's not required however, and if you would prefer it was dropped I'll get rid of it.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -91,42 +126,54 @@ child-binding:
-          - 1
-          - 2
-          - 3
+          - "mhz-50"
+          - "mhz-100"
+          - "mhz-150"
+          - "mhz-200"
```
suggestion: `50-mhz` (as it would read on a text)

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -163,6 +157,18 @@ child-binding:
           Select slew rate for pin. Corresponds to SRE field in IOMUXC peripheral
           0 SRE_0_Slow_Slew_Rate — Slow Slew Rate
           1 SRE_1_Fast_Slew_Rate — Fast Slew Rate
+      nxp,pull:
+        required: true
+        type: string
+        enum:
+          - "no-pull"
+          - "pull-enable"
+        description: |
+          Select pull enable bit. If this bit is selected, than the pin
+          pull/keeper settings take effect. If it is cleared, no pull is applied
+          to the pin.
+          0 PKE_Disable- No pull applied to pin
+          1 PKE_Enable- pull or keeper applied to pin depending on bias,bus-hold
```
you could infer this from the existence of `bias-pull-up`, `bias-pull-down` or `bias-bus-hold`, right?

```
pke = DT_PROP(node_id, bias_pull_up) | DT_PROP(node_id, bias_pull_down) | DT_PROP(node_id, bias_bus_hold);
```

danieldegrasse replied:
I could, and originally took this approach, but the concern with inferring the value of `pke` was the same issue I outlined in [this comment below](https://github.com/zephyrproject-rtos/zephyr/pull/42238#issuecomment-1031255945), namely that the "default" value for `PKE` is almost always 1, so the user might end up with unexpected affects if they level all boolean properties unselected on a pin. My plan right now is to drop `bias-bus-hold` and only set PKE to 0 if `bias-disable` is selected.

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi:
```diff
@@ -2,57 +2,66 @@
-		iomuxc_gpio_ad_b0_13_lpuart1_rx: iomuxc_gpio_ad_b0_13_lpuart1_rx {
-			pinmux = <IOMUXC_GPIO_AD_B0_13_LPUART1_RX>;
+		group0 {
+			mux = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX_MUX>,
+				<IOMUXC_GPIO_AD_B0_13_LPUART1_RX_MUX>;
+			mode = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX_MODE>,
+				<IOMUXC_GPIO_AD_B0_13_LPUART1_RX_MODE>;
+			input = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX_INPUT>,
+				<IOMUXC_GPIO_AD_B0_13_LPUART1_RX_INPUT>;
+			daisy = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX_DAISY>,
+				<IOMUXC_GPIO_AD_B0_13_LPUART1_RX_DAISY>;
+			cfg = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX_CFG>,
+				<IOMUXC_GPIO_AD_B0_13_LPUART1_RX_CFG>;
```
this is definitely a solution to the 32-bit limitation. What you mention of using nodes containing all these fields in a single place is also a good solution (less verbose, so probably worth giving it a try).

gmarull commented:
> @gmarull I've updated the PR to use string `enums`, per your suggestion. For the pin mux properties, We've prefer to avoid trying to pack all the data into a 32 bit field, primarily for scalability concerns. Some of our higher pin count Linux cores have a companion M4 core, and use the same IOMUXC peripheral. The approach I used in this update was to split pin configuration fields across several properties. The other option I considered was a method using DTS phandles to define the pin mux options, and using those phandles in pin groups, like so:
> 
> soc `dtsi` file:
> 
> ```
> &iomuxc {
>         iomuxc_gpio_ad_b0_12_lpuart1_tx: IOMUXC_GPIO_AD_B0_12_LPUART1_TX {
>         	pinmux = <0x401f80ec 2 0x0 0 0x401f82dc>;
>         };
>         
>         ...
>         
>         iomuxc_gpio_ad_b0_13_lpuart1_rx: IOMUXC_GPIO_AD_B0_13_LPUART1_RX {
>         	pinmux = <0x401f80f0 2 0x0 0 0x401f82e0>;
>         };
> };
> ```
> 
> board `dtsi` file
> 
> ```
> &pinctrl {
>     lpuart1_default: lpuart1_default {
> 		group0 {
> 			pinmux = <&iomuxc_gpio_ad_b0_12_lpuart1_tx>,
> 				<&iomuxc_gpio_ad_b0_13_lpuart1_rx>;
> 			nxp,pull = "pull-enable";
> 			bias-bus-hold;
> 			drive-strength = "r0-6";
> 			slew-rate = "slow";
> 			nxp,speed = "mhz-100";
> 		};
> 	};
> }
> ```
> 
> My primary concern with this approach is that it mixes both pin node and pin group approaches, and a user might incorrectly assume they could use the `iomuxc` nodes as pinctrl nodes directly.

As soon as it is well documented, I don't see a problem, really. It's less verbose than the current approach, so it's probably worth giving it a try. I have never tried this model, not sure about the macro hackery required (or if doable) :-) 
> 
> The other change in the latest push is a new property, `nxp,pull`. This property sets the `PKE` bit in the `IOMUXC` peripheral. The rationale was that if a user did not provide either `bias-bus-hold`, `bias-pull-up`, or `bias-pull-down` as a property, the pin would have all pull up and keeper functions disabled. Although this is a valid configuration, the typical "default" configuration for a pin that is not being pulled up on an RT SOC would be to enable the keeper.

You may introduce the negated `nxp,pke-disable` property so that you always enable `PKE` unless explicitly disabled.

> 
> Adding this property forces a user to understand that if they don't supply bus bias configuration to the pinctrl driver the pin will have its keeper disabled. The other option I considered here is to default to `PKE=1, PUE=0` if the user does not supply `bias-disable` as a property

danieldegrasse commented:
@gmarull, any feedback since the latest push? I've switched to using DTS nodes to encode pinmux settings, and dropped `bias-bus-hold` in favor of `bias-disable` to explicitly disable the keeper, so that the "default" for a pin has the keeper set to enabled.

gmarull commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -226,3 +231,11 @@ zephyr_udc0: &usb1 {
+	status = "okay";
+};
+
+&pinctrl {
+	status = "okay";
+};
```
nit: since pinctrl is required by most (if not all) applications, you may enable this by default

danieldegrasse replied:
Fixed in latest push

gmarull commented on drivers/pinctrl/Kconfig.mcux:
```diff
@@ -0,0 +1,9 @@
+config PINCTRL_MCUX_RT
+	bool "Pin controller driver for MCUX RT1xxx MCUs"
+	depends on SOC_SERIES_IMX_RT
+	default y
```
nit: default based on `dt_compat_status`, see other drivers

danieldegrasse replied:
Fixed in latest push

gmarull commented on drivers/pinctrl/pinctrl_mcux_rt.c:
```diff
@@ -0,0 +1,47 @@
+#include <fsl_iomuxc.h>
+#include <fsl_gpio.h>
+
+#define DT_DRV_COMPAT nxp_mcux_rt_pinctrl
```
nit: by convention move this to the top

danieldegrasse replied:
Fixed in latest push

gmarull commented on drivers/pinctrl/pinctrl_mcux_rt.c:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2021 NXP
```
2022 :-)?

danieldegrasse replied:
Thanks for noticing, fixed this and a few other copyrights for new files

gmarull commented on drivers/pinctrl/pinctrl_mcux_rt.c:
```diff
@@ -0,0 +1,47 @@
+	return 0;
+}
+
+SYS_INIT(mcux_pinctrl_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
note: make sure this happens before other PRE_KERNEL_1 devices that use pinctrl. For example UART tends to be initialized early.

gmarull reviewed (commented):
LGTM, I think current approach keeps things simple. Just added a few nitpicks, will do a last pass review once ready (looks like some commits need to be squashed/fixed up).

gmarull reviewed (commented):
LGTM, a few comments from my side.

danieldegrasse commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,153 @@
+      bias-pull-up-value:
+        required: false
+        type: string
+        default: "47k"
```
Although the SoC does not have a consistent default value across all pins, 47k is the most common reset value, so I selected it as the default. I'll add this to the property description.

danieldegrasse commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -456,3 +476,37 @@ static const struct mcux_lpuart_config mcux_lpuart_##n##_config = {	\
+	.flow_ctrl = DT_INST_PROP(n, hw_flow_control) ?	\
+		UART_CFG_FLOW_CTRL_RTS_CTS : UART_CFG_FLOW_CTRL_NONE,\
+	IRQ_FUNC_INIT							\
+}
```
That's the right solution, I believe I went with this option initially because the existing definitions were already duplicated. I've cleaned up the device definitions in the LPUART driver, and the latest push uses conditional sub-defines

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+    description: |
+      MCUX RT pin controller pin configuration node.
+    properties:
+      pins:
```
nit: shouldn't this be `pinmux`?

gmarull commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -335,8 +340,19 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
 	}
+}
+
+#ifdef CONFIG_PINCTRL
+static void mcux_rt_sai_pin_init(const struct i2s_mcux_config *dev_cfg)
+{
+	int err;
 
+	err = pinctrl_apply_state(dev_cfg->pinctrl, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		LOG_ERR("mclk pinctrl setup failed (%d)", err);
+	}
 }
+#endif
```
nit: please remove this wrapper, has no value and hides pinctrl.

gmarull commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+  for example, here is an group configuring LPUART1 pins:
+
+  group0 {
+    pins = <&iomuxc_gpio_ad_b0_12_lpuart1_tx,
```
last nitpick: `pinmux`

danieldegrasse replied:
Fixed in latest push. Thank you for working through this PR with me, I know it's taken a lot of your time but I think the code has improved immensely since the initial revision, and I'm really happy with the final result.

tejlmand reviewed (dismissed):
lgtm

dleach02 commented:
I've merged the HAL so good to fix up west.yaml

danieldegrasse commented on soc/arm/nxp_imx/rt/pinctrl_soc.h:
```diff
@@ -0,0 +1,84 @@
+
+#define Z_PINCTRL_MCUX_RT_PINCFG_INIT(node_id)							\
+	((DT_PROP(node_id, input_schmitt_enable) << MCUX_RT_INPUT_SCHMITT_ENABLE_SHIFT) |	\
+	IF_ENABLED(DT_PROP(node_id, bias_pull_up), (DT_ENUM_IDX(node_id, bias_pull_up_value)	\
```
@dleach02 should be good to merge now. The issue was that I was using a bitwise AND here, but the enum values for `bias_pull_up` / `bias_pull_down` can range from 0-3.

tejlmand reviewed (dismissed):
lgtm - re-approved

carlescufi commented:
@dleach02 ping

henrikbrixandersen commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+      &iomuxc_gpio_ad_b0_13_lpuart1_rx>;
+    drive-strength = "r0-6";
+    slew-rate = "slow";
+    nxp,speed = "mhz-100";
```
```suggestion
    nxp,speed = "100-mhz";
```

danieldegrasse replied:
Fixed in latest push

henrikbrixandersen commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+          the drive strength is expressed as a output impedance at a given voltage,
+          but maximum current values can be calculated from these impedances
+          for a specific load impedance.
+          000 DSE_0_output_driver_disabled_ — output driver disabled
+          001 DSE_1_R0_1 — 157 Ohm impedance @3.3V, 260 Ohm impedance @1.8V
+          010 DSE_2_R0_2 — 78 Ohm @3.3V, 130 Ohm @1.8V
+          011 DSE_3_R0_3 — 53 Ohm @3.3V, 88 Ohm @1.8V
+          100 DSE_4_R0_4 — 39 Ohm @3.3V, 65 Ohm @1.8V
+          101 DSE_5_R0_5 — 32 Ohm @3.3V, 52 Ohm @1.8V
+          110 DSE_6_R0_6 — 32 Ohm @3.3V, 43 Ohm @1.8V
+          111 DSE_7_R0_7 — 26 Ohm @3.3V, 37 Ohm @1.8V
```
These values (`000` to `111`) should likely be removed since they are set using an enum.

danieldegrasse replied:
I personally think they are useful to have- the driver uses the enum IDX as the value to write to the register, and these values tell you exactly what will be written to the `DSE` register. That way a user looking at this file does not have to guess what effect `drive-strength="r0"` will have at the register level, or look into the driver code itself.

henrikbrixandersen commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+          Corresponds to the PUS field in the IOMUXC peripheral.
+          47k resistor selected as default due to this being the default pullup
+          value on most SOC pins
+          00 Unused- no change will be applied to pin
+          01 PUS_1_47K_Ohm_Pull_Up — 47K Ohm Pull Up
+          10 PUS_2_100K_Ohm_Pull_Up — 100K Ohm Pull Up
+          11 PUS_2_22K_Ohm_Pull_Up — 22K Ohm Pull Up
```
These values (`00` to `11`) should likely be removed since they are set using an enum.

danieldegrasse replied:
See above comment

henrikbrixandersen commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+          - "fast"
+        description: |
+          Select slew rate for pin. Corresponds to SRE field in IOMUXC peripheral
+          0 SRE_0_Slow_Slew_Rate — Slow Slew Rate
+          1 SRE_1_Fast_Slew_Rate — Fast Slew Rate
```
These values (`0` to `1`) should likely be removed since they are set using an enum.

danieldegrasse replied:
See above comment

henrikbrixandersen commented on dts/bindings/pinctrl/nxp,mcux-rt-pinctrl.yaml:
```diff
@@ -0,0 +1,156 @@
+          - "200-mhz"
+        description: |
+          Sets pin speed. Corresponds to SPEED field in IOMUXC peripheral
+          00 SPEED_0_low_50MHz_ — low(50MHz)
+          01 SPEED_1_medium_100MHz_ — medium(100MHz)
+          10 SPEED_2_medium_150MHz_ — medium(150MHz)
+          11 SPEED_3_max_200MHz_ — max(200MHz)
```
These values (`00` to `11`) should likely be removed since they are set using an enum.

danieldegrasse replied:
See above comment

henrikbrixandersen reviewed (changes requested):
Seeing that you add pinctrl support to the lpuart driver, I tried testing this on the `twr_ke18f` board (`west build -b twr_ke18f samples/hello_world/ -- -DCONFIG_PINCTRL=y`), which uses the Kinetis pinctrl driver along with the lpuart driver - but compilation of the lpuart driver is now failing (see #43486).

danieldegrasse commented:
> Seeing that you add pinctrl support to the lpuart driver, I tried testing this on the twr_ke18f board
@henrikbrixandersen, just to confirm: #43486 fixed pinctrl support on the `twr_ke18f`, and the lpuart driver now works with pinctrl, correct? Just want to verify that there is not an issue with this PR, and that the bug was in the Kinetis pinctrl implementation

henrikbrixandersen commented:
> @henrikbrixandersen, just to confirm: #43486 fixed pinctrl support on the `twr_ke18f`, and the lpuart driver now works with pinctrl, correct? Just want to verify that there is not an issue with this PR, and that the bug was in the Kinetis pinctrl implementation

@danieldegrasse Sorry, I somehow missed your question. Yes, the problem was in the Kinetis pinctrl implementation and was fixed by #43486.
